blog: add battery life instrumentation post (#201)
tailscale-dev/tailscale-dev: data/blog/battery-life.mdx

---
title: 'Battery life instrumentation'
date: '2023-04-25'
tags:
  - 'oss'
  - 'mobile'
authors: ['mihai']
images: ['/images/battery-life/social.png']

> summary: 'Instrumentation that we’re adding to the Tailscale clients to have a scalable approach to battery life investigations.'

Reviewer: For consistency's sake, I'd prefer this to be more of a full sentence. Maybe:

We've added instrumentation to the Tailscale clients to make our investigations into batter life issues more scalable.

Or something like that?

To get a handle on this situation, we can make some simplifying assumptions. As seen in the graphs above (“BB” in power logs refers to [baseband radio](https://en.wikipedia.org/wiki/Baseband_processor)), the main contributor to Tailscale’s battery use is how often it uses a phone’s cellular connection. While there is some CPU overhead from WireGuard encryption and [checksumming](https://tailscale.com/blog/more-throughput/#tun-checksum-offload), it has not been a significant factor in mobile usage. Therefore we can use network traffic as a proxy for overall battery use. Specifically, we want to know how often we’re keeping the cellular radio in a [high-power state](https://developer.android.com/training/connectivity/network-access-optimization#radio-state) — there’s a big difference between sending 1,000 bytes in one go, or in 1,000 1-byte packets, one every second (for a total of 16 minutes).

> The problem thus becomes “how do we measure our own network use?” We first attempted to go [Darwin spelunking](https://bitsplitting.org/2023/03/17/spelunking-apples-open-source/) - hoping that the [network statistics](http://newosxbook.com/bonus/vol1ch16.html#:~:text=private%20NetworkStatistics%20framework) that the OS keeps would allow us to get this information in a low-overhead way. However, the APIs we wanted were either not available on all the platforms we wanted, or were too coarse-grained (without knowing which connection they’re tracking, we can’t distinguish between different kinds of traffic).

Reviewer: em dash after spelunking: —

---
title: 'Battery life instrumentation'
date: '2023-04-25'
tags:
  - 'oss'
  - 'mobile'
authors: ['mihai']
images: ['/images/battery-life/social.png']
summary: 'Instrumentation that we’re adding to the Tailscale clients to have a scalable approach to battery life investigations.'
---

> Battery consumption is one of the top areas where we’ve head that Tailscale needs to improve, via both [GitHub](https://github.com/tailscale/tailscale/issues/3363) and [social media](https://twitter.com/search?q=tailscale%20battery&src=typed_query&f=live). There’s always been [work on battery-related issues](https://github.com/tailscale/tailscale/issues?q=label%3Abattery+is%3Aclosed), but it’s not something that can ever be declared “done”:

Reviewer 2: Heard

> Battery consumption is one of the top areas where we’ve head that Tailscale needs to improve, via both [GitHub](https://github.com/tailscale/tailscale/issues/3363) and [social media](https://twitter.com/search?q=tailscale%20battery&src=typed_query&f=live). There’s always been [work on battery-related issues](https://github.com/tailscale/tailscale/issues?q=label%3Abattery+is%3Aclosed), but it’s not something that can ever be declared “done”:

Reviewer 3: ```suggestion
Battery consumption is one of the top areas where we’ve heard that Tailscale needs to improve, via both [GitHub](https://github.com/tailscale/tailscale/issues/3363) and [social media](https://twitter.com/search?q=tailscale%20battery&src=typed_query&f=live). Tailscale has continuously [worked on battery-related issues](https://github.com/tailscale/tailscale/issues?q=label%3Abattery+is%3Aclosed), but it’s not something that can ever be declared “done”, for a few reasons:
```

> - Tailscale’s feature set evolves — using the [NextDNS integration](https://tailscale.com/blog/nextdns/) also means that Tailscale is now responsible for all DNS requests issued by a device.

Reviewer 3: ```suggestion
- Tailscale’s feature set evolves. For example, using the [NextDNS integration](https://tailscale.com/blog/nextdns/) means that Tailscale is now responsible for all DNS requests issued by a device.
```
Not sure I get this - weren't we already responsible with MagicDNS? Is the point that it's routing more DNS queries off the device?

Author: Good point. Though MagicDNS only went GA a few months ago, and was not enabled by default before then. Changed it to incorporate that too.

- The operating system changes, and Apple or Google may intentionally or [unintentionally](https://github.com/tailscale/tailscale/issues/4646#issuecomment-1257251332) change how the application operates and when it’s allowed to run, and how much power it consumes.

> - New patterns of user behavior may arise, for example there may new reasons to [use an exit node](https://www.youtube.com/watch?v=9CunwUs08og) more frequently.

Reviewer 4: we might not want to link to this from an official Tailscale blog post. 😄

> - New patterns of user behavior may arise, for example there may new reasons to [use an exit node](https://www.youtube.com/watch?v=9CunwUs08og) more frequently.

Reviewer 3: Agreed.

Reviewer 4: (sorry, accidentally deleted comment so this is missing context now.  Was just saying I'm not sure we necessarily want to link to this)

> - New patterns of user behavior may arise, for example there may new reasons to [use an exit node](https://www.youtube.com/watch?v=9CunwUs08og) more frequently.

Reviewer 3: ```suggestion
- New patterns of user behavior may arise. Users could starting using Tailscale more often as more of their devices are available via Tailscale, or use Tailscale exit nodes more frequently to access certain resources.
```
super vague, I'm sure this could be worded better

> One of the challenges in this work is that it’s hard to get aggregated and actionable data. It’s great that recent versions of iOS and Xcode will collect and aggregate [metrics about battery use](https://developer.apple.com/documentation/xcode/analyzing-your-app-s-battery-use), but the numbers are delayed and very coarse-grained. In the graph below, it’s great that Tailscale 1.36 showed a marked improvement, but it’s not clear why (perhaps it was due to the [interface binding changes](https://tailscale.dev/blog/darwin-spelunking), or perhaps it was something else). If the direction had been reversed, we would be hard-pressed to know where to begin investigating.

Reviewer 3: ```suggestion
One of the challenges in this work is that it’s hard to get aggregated and actionable data. It’s great that recent versions of iOS and Xcode will collect and aggregate [metrics about battery use](https://developer.apple.com/documentation/xcode/analyzing-your-app-s-battery-use), but the numbers are delayed and very coarse-grained. In the graph below, Tailscale v1.36 showed a marked improvement compared to v1.34, but it’s not clear why (perhaps it was due to the [interface binding changes](https://tailscale.dev/blog/darwin-spelunking), or perhaps it was something else). If the direction had been reversed, we would be hard-pressed to know where to begin investigating.
```
Did we do something specific in 1.36? It sounds like we are more pleasantly surprised than did something to fix it.

Author: We did do something (the interface binding changes), but we were not confident that they would show up in the graphs (and have no way to be sure, since they're so coarse-grained). Tweaked the text a bit.

<figure className="float-none text-center">
  <Image
    src="/images/battery-life/ios-background-battery-use.png"
    width={862}
    height={430}
    alt="Battery use as reported by Xcode. Tailscale 1.36 shows a 40% drop."
  />
</figure>

> On the opposite end of the aggregation spectrum, we also [receive user reports](https://github.com/tailscale/tailscale/issues/3363#issuecomment-1297072512) based on-device statistics. While we’re able to correlated observed behavior with a specific user’s tailnet configuration, it’s somewhat frustrating to piece together things from a screenshot of the “Battery” pane of the iOS Settings app. To get more fine-grained data, we’ve been inspired by [Saagar Jha](https://twitter.com/_saagarjha/status/1528937319171887104)’s [Effective Power](https://github.com/saagarjha/EffectivePower) tool to [build some better visualizations](https://observablehq.com/d/d4955805772b34be) for on-device power logs:

Reviewer 3: ```suggestion
On the opposite end of the aggregation spectrum, we also [receive user reports](https://github.com/tailscale/tailscale/issues/3363#issuecomment-1297072512) based on-device statistics. While we’re able to correlated observed behavior with a specific user’s tailnet configuration, there's little we can do from just a screenshot of the “Battery” pane of the iOS Settings app. To get more fine-grained data, we’ve been inspired by [Saagar Jha](https://twitter.com/_saagarjha/status/1528937319171887104)’s [Effective Power](https://github.com/saagarjha/EffectivePower) tool to [build some better visualizations](https://observablehq.com/d/d4955805772b34be) for on-device power logs:
```

- Control plane: communication with the [coordination server](https://tailscale.com/blog/how-tailscale-works/#the-control-plane-key-exchange-and-coordination)
- DNS: both UDP and DNS-over-HTTPS
- Port-mapping: probes done as part of [NAT traversal](https://tailscale.com/blog/how-nat-traversal-works/)
- Logging: diagnostics and metrics uploaded by the client

This instrumentation allows us to do two things:

1. [Derive the cellular radio power state](https://github.com/tailscale/tailscale/pull/7777): while it’s a very simple model (assuming a 3G radio and that Tailscale is the only thing generating traffic), it allows us to have a single number that we can aggregate easily.
2. [Logging](https://github.com/tailscale/tailscale/pull/7498): we can keep track of the type of network traffic done by a client with high-precision. These logs are somewhat expensive to maintain and upload, thus it’s only done for a subset of users. They also very detailed, thus building visualization tooling is key.

With this instrumentation in place, we can have a [workflow](https://en.wikipedia.org/wiki/OODA_loop) for tackling battery life investigations:

> 1. Collect high-level aggregated statistics (e.g. the fraction of time spent in high-power states), broken down by Tailscale version

Reviewer 3: ```suggestion
1. Collect high-level aggregated statistics (e.g. the fraction of time spent in high-power states), broken down by Tailscale version.
```

2. If a new version or cohort of clients shows a significant change, pick representative clients and look at their network usage.
3. Based on the observations, make changes to the clients (or possibly change the instrumentation).
4. Deploy changes, and see if they have the desired effect.

> All of this instrumentation has landed relatively recently, but we’ve already been able to use it to investigate [some](https://github.com/tailscale/tailscale/issues/7617) [issues](https://github.com/tailscale/tailscale/issues/7850):

Reviewer 3: ```suggestion
All of this instrumentation has landed relatively recently, but we’ve already been able to use it to identify and begin investigating [some](https://github.com/tailscale/tailscale/issues/7617) [issues](https://github.com/tailscale/tailscale/issues/7850):
```

<figure className="float-none text-center">
  <Image
    src="/images/battery-life/sockstats-tx.png"
    width="1021"
    height="370"
    alt="Graph showing WireGuard handshake attempts every 5 seconds."
  />
  <figcaption>
    _If you keep sending packets every 5 seconds, you’re going to have a bad time_
  </figcaption>
</figure>

> Tech industry blog posts stereotypically end with “We rewrote our app in language X/ported it to framework Y/adopted methodology Z and all our problems were solved!”. It would be nice if all this work would lead us to a single root cause that would let us say “Tailscale’s battery life is negligible now.” However, based on our experience thus far, the much more likely outcome is that there will be many small issues or occasional interactions between sub-systems that lead to excessive battery use. It’s also likely that we’ll need to improve our tooling as we discover new things. The hope is that we’ve now established the infrastructure to be able to investigate such issues more easily, as well as keep an eye out for any regressions.

Reviewer 3: ```suggestion
Tech industry blog posts stereotypically end with “We rewrote our app in language X/ported it to framework Y/adopted methodology Z and all our problems were solved!”. Unfortunately, we're not there. It would be nice if all this work would lead us to a single root cause that would let us say “Tailscale’s battery life is negligible now.” Based on our experience thus far, the much more likely outcome is that there will be many small issues or occasional interactions between sub-systems that lead to excessive battery use. It’s also likely that we’ll need to improve our tooling as we discover new things.

We’ve now established the infrastructure to be able to investigate battery issues more easily, as well as keep an eye out for any regressions. The hope now is that we'll be able to see noticeable improvements as we tackle these issues.
```

Author: "Unfortunately, we're not there." isn't quite the right sentiment -- I think such absolutist conclusions are wrong/marketing fluff. Tweaked the text a bit.
